tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/ExemplarMetadataProvider.java

```diff
@@ -0,0 +1,23 @@
+/*
+ * (c) Copyright 2024 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.tritium.metrics.registry;
+
+import java.util.Optional;
+
+public interface ExemplarMetadataProvider<T> {
+    Optional<T> collect();
```

Comment:
We may want to use something like this instead, since it can potentially be a hot path. 

```
@Nullable
T collect();`
```

When the optional is empty, that's relatively inexpensive, but the additional object wrapper allocation could impact us if we're not careful.
We generally try to avoid using nulls, but we can make trade-offs in hot code-paths, and this isn't an API that users will be interacting with, so it may be safe.
I don't have a terribly strong opinion on this one, mostly thinking out loud.

Reply:
The other piece we should think about is how we plan to expose the measured value. When a histogram is updated with value `123`, we don't currently expose a way for the exemplar to reference that value in any way.

It may be helpful to include that value in `collect(long value);`. With only that, the exemplar samples will need to duplicate the value already held in the reservoir, which might be reasonable, but we could instead consider exposing something different from `ExemplarsCapture` like 

```diff
- <T> List<T> getSamples(ExemplarMetadataProvider<T> provider);
+ <T> List<LongExemplar<T>> getSamples(ExemplarMetadataProvider<T> provider);
```
Where `LongExemplar<T>` is roughly
```java
interface LongExemplar<T> {
    T exemplar();
    long value();
}
```

Reply:
Yeah good point! I'm definitely more comfortable not duplicating the samples at the reservoir level - so made the switch you proposed. I do have the values duplicated in the Snapshot for easier implementation but that should be way less of a concern - let me know if you disagree.